snowplow_tracker/emitters.py

```diff
@@ -187,74 +202,76 @@ def reached_limit(self) -> bool:
         :rtype: bool
         """
         if self.byte_limit is None:
-            return len(self.buffer) >= self.buffer_size
+            return len(self.buffer) >= self.batch_size
         else:
             return (self.bytes_queued or 0) >= self.byte_limit or len(
                 self.buffer
-            ) >= self.buffer_size
+            ) >= self.batch_size
 
     def flush(self) -> None:
         """
         Sends all events in the buffer to the collector.
         """
-        with self.lock:
-            self.send_events(self.buffer)
+        if self.retry_delay > 0:
```

Comment:
I think we need to acquire the `lock` here in before checking the `retry_delay` because there might be an ongoing send operation in which the delay could be modified. So I would change this to something like:

```python
with self.lock:
    delay = self.retry_delay

if delay > 0:
    self.set_flush_timer(delay)
else:
    self._flush_now()
```